sql/connect/server/src/main/scala/org/apache/spark/sql/connect/ml/MLHandler.scala

```diff
@@ -125,6 +127,15 @@ private[connect] object MLHandler extends Logging {
         val dataset = MLUtils.parseRelationProto(fitCmd.getDataset, sessionHolder)
         val estimator =
           MLUtils.getEstimator(sessionHolder, estimatorProto, Some(fitCmd.getParams))
+
+        // pre-training model size check
+        val maxSize = conf.getConf(Connect.CONNECT_SESSION_ML_CACHE_SINGLE_ITEM_SIZE)
+        if (maxSize > 0) {
+          val estimatedSize = estimator.estimateModelSize(dataset)
```

Comment:
How accurate is this? How much state is added to the model while fitting?

Reply:
That depends, we are trying to estimate an upper bound of this size. E.g.

> - Given a LogisticRegression estimator, assume the coefficients are dense, even though the actual fitted model might be sparse (by L1 penalty).
> - Given a tree model, assume all underlying trees are complete binary trees, even though some branches might be pruned or truncated.

Reply:
we will need unit tests to guard this estimation